### DISCUSSION TITLE
请问模型训练好之后怎么应用？

### DISCUSSION THREAD

**Discussion by HickeyHsu** [Congyuwang/AlphaNetV3#13] [Q&A] [answered]
十分感谢分享！我用自己的数据训练了一个模型，请问怎么应用训练好的模型？（即测试集应用）
即：如何给模型输入测试集序列，输出return的label？

我尝试model.predict(test_data),其中test_data是一个TimeSeriesData，但似乎并不能成功。

`    # load entire model using load_model() from alphanet module
    model = load_model("model.bt")

    # only load weights by first creating a model instance
    model = AlphaNetV3(l2=0.001, dropout=0.0)
    model.load_weights("weights.bt")
    test_data=TimeSeriesData(
        dates=test_df["int_date"].values,                   # date column            
        data=test_df[metrics_std].values,                # data columns
        labels=test_df["future_return"].values # label column
    )
    output=model.predict(test_data.data)
    print(output)`

**Comment by Congyuwang** [comment]
您可以尝试将timeseries data传给trainval data, 设置train length为0（无训练长度），设置合适的val length。使用get获取tensorflow data，get的返回值可以用于predict

**Reply by HickeyHsu** [reply]
我尝试了，发现数据检查里要求history_length不能大于train length也不能小于1

**Comment by Congyuwang** [comment]
或者你也可以尝试自己传递numpy.array给predict函数。

**Comment by Congyuwang** [comment]
不需要用`TimeSeriesData`。你可以直接从`test_df[metrics_std].values`取出相应的历史长度(比如30天或者15天，取决于你的模型)，比如。也不需要用`predict`，直接call `model`：

```python
# 前15天
data = test_df[metrics_std].values[:15, :]
prediction = model(data)
```
如果可以的话，告诉我一下

**Reply by HickeyHsu** [reply]
尝试了，似乎还是不行，提示：
`IndexError : list index out of range`
我的history_length设置的20,8个特征；看了一下data的shape是（20,8）应该没错。
我将我的数据和代码放到了这里：https://github.com/HickeyHsu/ecQuant4CB；如果您有空的话麻烦帮我看看 谢谢！

**Reply by Congyuwang** [reply]
下面这个回复能用吗？

**Reply by HickeyHsu** [reply]
抱歉 我后来外出办事了，明天尝试一下

**Comment by Congyuwang** [comment, marked as answer]
Try this:

```python
model(test_df[metrics_std].values[30:50, :].reshape(-1, 20, 8))
```

Two things to notice:
- Shouldn't contain NAN in NN prediction input
- need a batch dimension, so do: `.rehspae(-1, x, x)`

**Reply by HickeyHsu** [reply]
刚刚简单测试了一下，是可以的，输出是一个值（1*1矩阵）。按照我对模型的理解，输出值应该是20天数据之后的下一天的“10日回报率估值”对吗？
另外想请问一下，训练数据每只标的最后10天的10日回报率标签都是未知的，这种情况下需要将最后10天数据剔除么？（我实验里没有剔除，也能正常运行）

**Reply by Congyuwang** [reply]
`[30:50]`不包含NA，如果输入包含NA，则输出NA。

**Reply by Congyuwang** [reply]
> 刚刚简单测试了一下，是可以的，输出是一个值（1*1矩阵）。按照我对模型的理解，输出值应该是20天数据之后的下一天的“10日回报率估值”对吗？

是否是下一天的10日回报率取决于训练时的标签是否是下一天的10日回报率。

**Reply by HickeyHsu** [reply]
谢谢回复~
我的意思是训练集的标签包含NA。
例如：我有一只股票的60日数据（所有特征都存在，不是NA），那么我计算10日回报率的时候，因为没有61-70日的价格数据，因此只有前50日可以算出10日回报率，51-60日是没有的标签的（但是有特征）。
那么51-60日的数据是否需要剔除之后再训练模型呢？目前我没有剔除这部分数据，依然可以运行训练，我不清楚这是否是合适的。

**Reply by Congyuwang** [reply]
> 谢谢回复~
> 
> 我的意思是训练集的标签包含NA。
> 
> 例如：我有一只股票的60日数据（所有特征都存在，不是NA），那么我计算10日回报率的时候，因为没有61-70日的价格数据，因此只有前50日可以算出10日回报率，51-60日是没有的标签的（但是有特征）。
> 
> 那么51-60日的数据是否需要剔除之后再训练模型呢？目前我没有剔除这部分数据，依然可以运行训练，我不清楚这是否是合适的。
> 
> 

训练集NA会被自动剔除。参考src/alphanet/data.py 第459行

**Reply by HickeyHsu** [reply]
好的，十分感谢！我去研究一下训练效果如何

**Reply by ReneeQRDu** [reply]
对需要的df_test做reshape之后，使用model.predict()也可以输出结果
直接调用model输出的是张量，用model.predict()输出的是ndarray。

想问一下，df_test.iloc[30:50,:]只要任意一行有NAN，最后返回的结果就会是NAN吗？

**Reply by Congyuwang** [reply]
你好是的。输入中有NA，输出一定为NA。
